libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,551 @@
+			}, layout.at(_offset));
+		}
+		int targetMultiplicity(size_t _offset)
+		{
+			if (!outputs.count(post.at(_offset)) && !generateSlotOnTheFly(post.at(_offset)))
+				return 0;
+			return multiplicity.at(post.at(_offset));
+		}
+		bool targetIsArbitrary(size_t _offset)
+		{
+			return _offset < post.size() && std::holds_alternative<JunkSlot>(post.at(_offset));
+		}
+		void swap(size_t _i)
+		{
+			yulAssert(!holds_alternative<PreviousSlot>(layout.at(layout.size() - _i - 1)) || !holds_alternative<PreviousSlot>(layout.back()), "");
+			std::swap(layout.at(layout.size() -  _i - 1), layout.back());
+		}
+		size_t sourceSize() { return layout.size(); }
+		size_t targetSize() { return post.size(); }
+		void pop() { layout.pop_back(); }
+		void pushOrDupTarget(size_t _offset) { layout.push_back(post.at(_offset)); }
+	};
+
+	Shuffler<ShuffleOperations>::shuffle(_layout, _post, _generateSlotOnTheFly);
+
+	// Now we can construct the ideal layout before the operation.
+	// "layout" has the declared variables in the desired position and
+	// for any PreviousSlot{x}, x yields the ideal place of the slot before the declaration.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, _layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	// This is a huge tradeoff between code size, gas cost and stack size.
+	auto generateSlotOnTheFly = [&](StackSlot const&) {
+		//return stack.size() > 12 && canBeFreelyGenerated(_slot);
+		// return canBeFreelyGenerated(_slot);
+		return false;
+	};
+
+	size_t previousLayoutSize = stack.size();
+	for (auto const& slot: stack)
+		if (util::findOffset(_operation.output, slot) || generateSlotOnTheFly(slot))
+				--previousLayoutSize;
+
+	auto layout = ranges::views::iota(0u, previousLayoutSize) |
+		ranges::views::transform([](size_t _index) { return PreviousSlot{_index}; }) |
+		ranges::to<vector<variant<PreviousSlot, StackSlot>>>;
+	// The call produces a known sequence of values.
+	layout += _operation.output;
+
+	stack = createIdealLayout(stack, layout, generateSlotOnTheFly);
+
+	if (auto const* assignment = get_if<CFG::Assignment>(&_operation.operation))
+		for (auto& stackSlot: stack)
+			if (auto const* varSlot = get_if<VariableSlot>(&stackSlot))
+				yulAssert(!util::findOffset(assignment->variables, *varSlot), "");
+
+	for (StackSlot const& input: _operation.input)
+		stack.emplace_back(input);
+
+	m_layout.operationEntryLayout[&_operation] = stack;
```

Comment:
Hm, I would not have expected this side-effect, to be honest.

Reply:
Also, why does it return something else than it stores?

Reply:
Hm... I mean it's only natural to store them while propagating through blocks... so maybe the function needs to be renamed.

Reply:
> Also, why does it return something else than it stores?

It stores the stack that is required immediately before the operation, while it returns a potentially compressed one that merely has to make it possible to create the one that's stored.

Reply:
I would say it's fine to document in the header what it stores in the member.